File: leetcode/875.ts
// Runtime: 154 ms, faster than 22.22% of TypeScript online submissions for Koko Eating Bananas.
// Memory Usage: 42 MB, less than 88.89% of TypeScript online submissions for Koko Eating Bananas.
// Binary serach seems to be popular in other submissions, I may be able to optomize some by not doing it as a recursive function.
function minEatingSpeed(piles: number[], h: number): number {
    let maxPile = 0;
    for(let i = 0; i < piles.length; i++){
        if(piles[i]>maxPile){
            maxPile = piles[i];
        }
    }
    return minEatingSpeedSearch(piles, h, maxPile, 1);
};

function minEatingSpeedSearch(piles: number[], h:number, hRangeHigh: number, hRangeLow: number){
    let count = 0;
    let rate = Math.floor((hRangeLow + hRangeHigh) / 2);
    if(hRangeHigh < hRangeLow){
        return Number.MAX_SAFE_INTEGER;
    }
    for(let i = 0; i < piles.length; i++){
        count += Math.ceil(piles[i]/rate);
        if(count > h){
            return minEatingSpeedSearch(piles, h, hRangeHigh, rate + 1);
        }
    }
    return Math.min(rate, minEatingSpeedSearch(piles, h, rate - 1, hRangeLow));
}

File: leetcode/875.cs
public class Solution {
    public int MinEatingSpeed(int[] piles, int h) {
        // we can binary search for the goal speed?
        int min = 1;
        int max = 1;
        for(int i = 0; i<piles.Length;i++){
            max = Math.Max(max, piles[i]);
        } 
        while(min < max){
            int mid = (min+max)/2;
            if(this.checkSpeed(piles, h, mid)){
                max = mid;
            }
            else{
                min = mid +1;
            }
        }
        return min;
    }
    public bool checkSpeed(int[] piles, int h, int speed){
        int hours = 0;
        for(int i = 0; i<piles.Length;i++){
            hours += piles[i]/speed;
            if(piles[i]%speed != 0){
                hours++;
            } 
        }
        return hours <= h;
    }
}
